core/consensus/babe/src/lib.rs

```diff
@@ -1190,6 +1192,117 @@ pub fn import_queue<B, E, Block: BlockT<Hash=H256>, I, RA, PRA>(
 	Ok((queue, timestamp_core, block_import, pruning_task))
 }
 
+/// Provide a list of authority keys that is current as of a given block.
+#[allow(deprecated)]
+fn authorities_at<C>(client: &C, at: &BlockId<<C as BlockOf>::Type>)
+	-> Result<Vec<AuthorityId>, ConsensusError>
+where
+	C: ProvideRuntimeApi + BlockOf + ProvideCache<<C as BlockOf>::Type>,
+	C::Api: BabeApi<<C as BlockOf>::Type>
+{
+	client
+		.cache()
+		.and_then(|cache| cache
+			.get_at(&well_known_cache_keys::AUTHORITIES, at)
+			.and_then(|v| Decode::decode(&mut &v[..]))
+		)
+		.or_else(|| BabeApi::epoch(&*client.runtime_api(), at).ok()
+			.map(|epoch| epoch.authorities.into_iter().map(|x| x.0).collect())
+		)
+		.ok_or_else(|| consensus_common::Error::InvalidAuthoritiesSet.into())
+}
+
+/// Provide the authority key, if any, that is controlled by this node as of the given block.
+fn authority<C>(client: &C, keystore: Arc<Store>) -> Option<AuthorityPair> where
+	C: ProvideRuntimeApi + BlockOf + ProvideCache<<C as BlockOf>::Type>
+	+ HeaderBackend<<C as BlockOf>::Type>,
+	C::Api: BabeApi<<C as BlockOf>::Type>
+{
+	let owned = keystore.public_keys::<AuthorityId>().ok()?;
+	let at = BlockId::Number(client.info().best_number);
+	// The list of authority keys that is current. By default this will just use the state of
+	// the best block, but you might want it to use some other block's state instead if it's
+	// more sophisticated. Grandpa, for example, will probably want to use the state of the last
+	// finalised block.
+	let authorities = authorities_at::<C>(client, &at).ok()?;
+	let maybe_pub = owned.into_iter()
+		.find(|i| authorities.contains(i));
+	maybe_pub.and_then(|public| keystore.key_pair(&public).ok())
+}
+
+/// Type of source for block sealing. Different consensus algorithms have different sealing
+/// methods; for PoW it'll be a miner or mining instance. For PoA/PoS it'll be a key type.
+pub trait SealingSource {
+	/// Human-readable description of this general class.
+	///
+	/// e.g. `"ProgPoW"`, `"S/R 25519 Key"`.
+	const SEALER_TYPE: &'static str;
+	/// Human-readable description of this specific instance.
+	///
+	/// e.g. `"OpenCL GPU Radeon 8970 @ slot 1"`, `"5Gav1mbbo348grBREHTYgevvf4Gfe5GFE4"`.
+	fn format(&self) -> String;
+}
+
+impl<T: app_crypto::AppPublic> SealingSource for T {
+	const SEALER_TYPE: &'static str = "Key";
+	fn format(&self) -> String { use primitives::crypto::Ss58Codec; self.to_ss58check() }
+}
+
+/// A runtime service for a consensus worker.
+///
+/// This is administered by the client service to help it spin up everything necessary. It can
+/// provide API hooks for handling user-level/RPC events. It can return information regarding
+/// the status of the service.
+pub trait Service<'a>: 'a {
```

Comment:
this and related stubs should be removed before merging - an issue with the pseudocode to go over options can contain it instead